docker/compose_files/docker-compose.yml

```diff
@@ -1,91 +1,111 @@
 version: '3.5'
 
-x-non-blocking-logging:
-  &non-blocking-logging
+x-non-blocking-logging: &non-blocking-logging
   driver: json-file
   options:
     mode: non-blocking
 
-x-codalab-env:
-  &codalab-env
-  - CODALAB_HOME=${CODALAB_SERVICE_HOME}
-  - CODALAB_USERNAME=${CODALAB_ROOT_USER}
-  - CODALAB_PASSWORD=${CODALAB_ROOT_PWD}
+x-codalab-env: &codalab-env
+  - CODALAB_HOME=${CODALAB_HOME}
+  - CODALAB_USERNAME=${CODALAB_USERNAME}
+  - CODALAB_PASSWORD=${CODALAB_PASSWORD}
+  - CODALAB_MYSQL_HOST=${CODALAB_MYSQL_HOST}
+  - CODALAB_MYSQL_PORT=${CODALAB_MYSQL_PORT}
+  - CODALAB_MYSQL_DATABASE=${CODALAB_MYSQL_DATABASE}
+  - CODALAB_MYSQL_USERNAME=${CODALAB_MYSQL_USERNAME}
+  - CODALAB_MYSQL_PASSWORD=${CODALAB_MYSQL_PASSWORD}
+  - CODALAB_ADMIN_EMAIL=${CODALAB_ADMIN_EMAIL}
+  - CODALAB_EMAIL_HOST=${CODALAB_EMAIL_HOST}
+  - CODALAB_EMAIL_USERNAME=${CODALAB_EMAIL_USERNAME}
+  - CODALAB_EMAIL_PASSWORD=${CODALAB_EMAIL_PASSWORD}
+
+# Properties that every service inherits.
+x-codalab-base: &codalab-base
+  environment:
+    *codalab-env
+  logging:
+    *non-blocking-logging
+  user: ${CODALAB_UID}
+  networks:
+    - service
+
+# We mostly want to run in user mode when we need to save files to disk.
+# But if we don't, easier to run as root.
+x-codalab-root: &codalab-root
+  user: 0:0
+
+# Services that need direct access to disk.
+x-codalab-server: &codalab-server
+  volumes:
+  - "${CODALAB_HOME}:${CODALAB_HOME}"
+  - "${CODALAB_BUNDLE_STORE}:${CODALAB_BUNDLE_STORE}"
+  - "${CODALAB_MONITOR_DIR}:${CODALAB_MONITOR_DIR}"
+
+############################################################
 
 services:
   rest-server:
     image: codalab/server:${CODALAB_VERSION}
     command: cl server
-    environment:
-      *codalab-env
-    networks:
-      - service
-    ports:
-      - ${CODALAB_REST_PORT}:2900
-    logging:
-      *non-blocking-logging
+    <<: *codalab-base
```

Comment:
What do the leading * and &'s mean in docker-compose files?

Reply:
It's a yaml thing to allow for modular definitions.